lib/miq_automation_engine/engine/miq_ae_engine/miq_ae_object_lookup.rb

```diff
@@ -0,0 +1,60 @@
+module MiqAeEngine
+  module MiqAeObjectLookup
+    def initialize_obj_entries
+      @lookup_table = []
+    end
+
+    def add_obj_entry(fqns, klass, instance, object)
+      parts = fqns.split('/')
+      domain = parts.shift
+      namespace = parts.join('/')
+
+      @lookup_table << {:domain    => domain.downcase,
+                        :namespace => namespace.downcase,
+                        :klass     => klass.downcase,
+                        :instance  => instance.downcase,
+                        :object    => object}
+    end
+
+    def find_obj_entry(path)
+      query = path[0] == '/' ? parse_obj_path(path[1..-1]) : parse_obj_path(path)
+
+      entries = @lookup_table.select { |entry| entry[:klass] == query['class_name'] }
```

Comment:
Since you are always looking up by entry[:klass], I think it would be better if the @lookup_table where a hash with the klass as the key (it can be part of the value too.  Then this line becomes an very efficient:

```ruby
@lookup_table[query['class_name']]
```

Reply:
@Fryguy The class can be in multiple domains.

Reply:
Still could use a lookup table, just store arrays by class